pkg/backend/metrics.go

```diff
@@ -0,0 +1,203 @@
+)
+
+const (
+	metricPackage = "backend.openshiftcluster"
+	metricInstall = metricPackage + ".installphase"
+	enabled       = "Enabled"
+	disabled      = "Disabled"
+	custom        = "Custom"
+	defaultSet    = "Default"
+	unknown       = "unknown"
+	empty         = "empty"
+)
+
+func (ocb *openShiftClusterBackend) emitMetrics(log *logrus.Entry, doc *api.OpenShiftClusterDocument, operationType, provisioningState api.ProvisioningState, backendErr error) {
+	dimensions := map[string]string{
+		"operationtype":     operationType.String(),
+		"provisioningstate": provisioningState.String(),
+		"resulttype":        string(ocb.getResultType(backendErr)),
+	}
+
+	ocb.gatherCorrelationID(doc, dimensions)
+	ocb.gatherMiscMetrics(doc, dimensions)
+	ocb.gatherAuthMetrics(doc, dimensions)
+	ocb.gatherNetworkMetrics(doc, dimensions)
+	ocb.gatherNodeMetrics(doc, dimensions)
+
+	ocb.logMetricDimensions(log, dimensions)
+	ocb.m.EmitGauge(metricInstall, 1, dimensions)
+}
+
+func (ocb *openShiftClusterBackend) getResultType(backendErr error) utillog.ResultType {
+	var resultType utillog.ResultType
+	err, ok := backendErr.(*api.CloudError)
+	if ok {
+		resultType = utillog.MapStatusCodeToResultType(err.StatusCode)
+	}
+	return resultType
+}
+
+func (ocb *openShiftClusterBackend) logMetricDimensions(log *logrus.Entry, dimensions map[string]string) {
+	for metric, value := range dimensions {
+		log.Info(fmt.Sprintf("%s: %s = %s", metricPackage, metric, value))
+	}
+}
+
+func (m *openShiftClusterBackend) gatherCorrelationID(doc *api.OpenShiftClusterDocument, dimensions map[string]string) {
+	if doc.CorrelationData != nil {
+		dimensions["correlationid"] = doc.CorrelationData.CorrelationID
+		dimensions["client_requestid"] = doc.CorrelationData.ClientRequestID
+		dimensions["requestid"] = doc.CorrelationData.RequestID
+		dimensions["async_operationsid"] = doc.AsyncOperationID
+	} else {
+		dimensions["correlationid"] = empty
+		dimensions["client_requestid"] = empty
+		dimensions["async_operationsid"] = empty
+	}
+}
+
+func (ocb *openShiftClusterBackend) gatherMiscMetrics(doc *api.OpenShiftClusterDocument, dimensions map[string]string) {
+	dimensions["subscriptionid"] = ocb.env.SubscriptionID()
+	dimensions["resourceid"] = doc.ResourceID
+	dimensions["location"] = doc.OpenShiftCluster.Location
+	dimensions["resourcegroup"] = doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID
+	dimensions["name"] = doc.OpenShiftCluster.Name
+	dimensions["ocpversion"] = doc.OpenShiftCluster.Properties.ClusterProfile.Version
+	dimensions["workercount"] = strconv.FormatInt(int64(len(doc.OpenShiftCluster.Properties.WorkerProfiles)), 10)
+	dimensions["rpversion"] = doc.OpenShiftCluster.Properties.ProvisionedBy
+
+	for flag, feature := range doc.OpenShiftCluster.Properties.OperatorFlags {
+		dimensions[fmt.Sprintf("operator-%s", flag)] = feature
+	}
+
+	if doc.OpenShiftCluster.Tags != nil {
+		dimensions["tags"] = enabled
+	} else {
+		dimensions["tags"] = disabled
+	}
+}
```

Comment:
Is this a useful dimension for us? We don't have any behavioral changes based on tags. I could see this being useful for us to e.g. identify clusters whose installs failed due to a policy requiring tags, but those are usually case-by-case situations where common metrics wouldn't help I don't think.

Reply:
We want to capture the use of all user provided input during cluster creation.  
So this isn't so much of a, will this help us see a problem, but rather are users using things? What things?